utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py

```diff
@@ -0,0 +1,360 @@
+            )
+
+        self.tensor_stats = {stat: TENSOR_STATS[stat] for stat in tensor_stats}
+
+        # Default hook registration
+        self.hm.register_hooks()
+
+    def _safe_tensor_stats(self, tensor: torch.Tensor) -> Dict[str, float]:
+        """
+        Safely compute tensor statistics with error handling.
+
+        Args:
+            tensor (torch.Tensor): Input tensor
+
+        Returns:
+            Dict of statistical metrics
+        """
+        stats = {}
+        for stat_name, stat_func in self.tensor_stats.items():
+            try:
+                stats[stat_name] = stat_func(tensor)
+            except Exception as e:
+                warnings.warn(f"Could not compute {stat_name} statistic: {e}")
+        return stats
+
+    def _track_metric(
+        self, metric_type: str, data: Dict[str, torch.Tensor], namespace: Optional[str] = None
+    ):
+        """Track metrics with enhanced statistics for a given metric type.
+
+        Args:
+            metric_type (str): Type of metric being tracked (activation/gradient/parameters)
+            data (Dict[str, torch.Tensor]): Dictionary mapping layer names to tensors
+            namespace (Optional[str]): Optional namespace to prefix the base namespace
+        """
+        # Construct the full namespace
+        full_namespace = f"{namespace}/{self.base_namespace}" if namespace else self.base_namespace
+
+        for layer, tensor in data.items():
+            if tensor is not None:
+                stats = self._safe_tensor_stats(tensor)
+                for stat_name, stat_value in stats.items():
+                    self.debug_metrics[f"{full_namespace}/{metric_type}/{layer}_{stat_name}"] = (
+                        stat_value
+                    )
+
+    def track_activations(self, namespace: Optional[str] = None):
+        """Track layer activations with enhanced statistics."""
+        activations = self.hm.get_activations()
+        self._track_metric("activation", activations, namespace)
+
+    def track_gradients(self, namespace: Optional[str] = None):
+        """Track layer gradients with enhanced statistics."""
+        gradients = self.hm.get_gradients()
+        self._track_metric("gradient", gradients, namespace)
+
+    def track_parameters(self, namespace: Optional[str] = None):
+        """Track model parameters with enhanced statistics."""
+        #TODO: Speed up for extracting parameters
+        with torch.no_grad():
+            parameters = {
+                name: param.grad
+                for name, param in self.model.named_parameters()
+                if param is not None and param.grad is not None
+            }
+            self._track_metric("parameters", parameters, namespace)
+
+    def watch(
+        self,
+        step: Union[int, float],
+        track_gradients: bool = True,
+        track_parameters: bool = True,
+        track_activations: bool = True,
+        namespace: Optional[str] = None,
```

Comment:
If this is acting as a `prefix` to the `base_namespace`, it should be named as such to make the information architeture clear

```suggestion
        prefix: Optional[str] = None,
```